### PR TITLE
Fix wayland client interaction

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

File: src/compositor/wayland/buffer.c [hunk 1]
```diff
@@ -334,3 +334,11 @@ end_access(
     wl_shm_buffer_end_access(shm_buffer);
 }
 
+void
+ws_wayland_buffer_release(
+    struct ws_wayland_buffer* self
+) {
+    struct wl_resource* res = ws_wayland_obj_get_wl_resource(&self->wl_obj);
```
Comment (matthiasbeyer) [inline src/compositor/wayland/buffer.c:348]: Could this possibly be `NULL`?

File: src/compositor/wayland/surface.c [hunk 2]
```diff
@@ -337,6 +338,14 @@ surface_frame_cb(
     uint32_t callback
 ) {
     //!< @todo: implement
+
+    struct ws_surface* surface = ws_surface_from_resource(resource);
+    surface->frame_callback = ws_wayland_client_create_resource(client,
+            &wl_callback_interface, 1, callback);
```
Comment (matthiasbeyer) [inline src/compositor/wayland/surface.c]: Please add more indentation here.

File: src/compositor/wayland/buffer.c (hunk 1 above)
Comment (TheNeikos) [inline src/compositor/wayland/buffer.c:348]: Yes, but what state should the application go if the buffer is NULL? There is no way to recover from that.

**Comment by neithernut** [comment]
I'll assume rhis is not WIP any more.

File: src/compositor/keyboard.c [hunk 3]
```diff
@@ -239,6 +239,31 @@ ws_keyboard_send_key(
...
+    uint32_t locked;
+    locked = xkb_state_serialize_mods(self->xkb->state, XKB_STATE_LOCKED);
+    uint32_t group;
+    group = xkb_state_serialize_layout(self->xkb->state,
+                                        XKB_STATE_LAYOUT_EFFECTIVE);
+    // ws_log(&log_ctx, LOG_DEBUG, "Direction : %d", direction);
+    // ws_log(&log_ctx, LOG_DEBUG, "Depressed : %x", depressed);
+    // ws_log(&log_ctx, LOG_DEBUG, "Latched : %x", latched);
+    // ws_log(&log_ctx, LOG_DEBUG, "Locked : %x", locked);
+    // ws_log(&log_ctx, LOG_DEBUG, "Group : %x", group);
```
Comment (neithernut) [inline src/compositor/keyboard.c]: Those should either be removed or made real calls (instead of comments).

Comment (TheNeikos) [inline src/compositor/keyboard.c]: You don't want them to be real calls, so I will remove them.

Comment (matthiasbeyer) [inline src/compositor/keyboard.c]: Maybe you can change them to `ws_dbg()` calls? Then, they will only be included in a debug build...

Comment (neithernut) [inline src/compositor/keyboard.c]: A removal would be preferrable.

**Comment by TheNeikos** [comment]
Okay, I fixed all the comments

File: src/compositor/keyboard.c [hunk 4]
```diff
@@ -239,6 +239,26 @@ ws_keyboard_send_key(
         return;
     }
 
+    enum xkb_key_direction direction;
+    direction = state == WL_KEYBOARD_KEY_STATE_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP;
```
Comment (matthiasbeyer) [inline src/compositor/keyboard.c]: Uuhh... I'm not sure about the binding here. Can you please add parentheses around the condition?

File: src/compositor/keyboard.c [hunk 5]
```diff
@@ -239,6 +239,26 @@ ws_keyboard_send_key(
...
+
+    uint32_t depressed;
+    depressed = xkb_state_serialize_mods(self->xkb->state, XKB_STATE_DEPRESSED);
+    uint32_t latched;
+    latched = xkb_state_serialize_mods(self->xkb->state, XKB_STATE_LATCHED);
+    uint32_t locked;
+    locked = xkb_state_serialize_mods(self->xkb->state, XKB_STATE_LOCKED);
+    uint32_t group;
+    group = xkb_state_serialize_layout(self->xkb->state,
+                                        XKB_STATE_LAYOUT_EFFECTIVE);
```
Comment (matthiasbeyer) [inline src/compositor/keyboard.c]: Can you please move the variable declarations up and the variable settings down and align them? Like so:

``` c
uint32_t ae;
uint32_t b;

ae = foo();
b  = bar();
```

Comment (matthiasbeyer) [inline src/compositor/keyboard.c]: (Would result in more readable code ... at least for me)

File: src/compositor/wayland/buffer.c [hunk 6]
```diff
@@ -120,6 +121,8 @@ end_access(
  *
  */
 
+static struct ws_logger_context log_ctx = { .prefix = "[Compositor/Surface]" };
```
Comment (matthiasbeyer) [inline src/compositor/wayland/buffer.c]: Missing whitespace after `]` in string!

**Comment by TheNeikos** [comment]
This has been changed again

**Comment by neithernut** [comment]
ACK.

**Pull request merged**